File: python/mirage/commands/builtin/history/find.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import fnmatch

from mirage.accessor.history import HistoryAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.utils.output import format_records
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.history.read import _VIEW_KEYS
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec

_VIEW_NAME = ".bash_history"


@command("find", resource="history", spec=SPECS["find"])
async def find(
    accessor: HistoryAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    name: str | None = None,
    type: str | None = None,
    iname: str | None = None,
    maxdepth: str | None = None,
    mindepth: str | None = None,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        return format_records([]), IOResult()
    p0 = paths[0]
    key = p0.strip_prefix if isinstance(p0, PathSpec) else p0
    if key.strip("/") not in _VIEW_KEYS:
        raise FileNotFoundError(key)
    if type == "d":
        return format_records([]), IOResult()
    if mindepth is not None and int(mindepth) > 0:
        return format_records([]), IOResult()
    matcher = iname or name
    candidate = _VIEW_NAME.lower() if iname else _VIEW_NAME
    pattern = matcher.lower() if (iname and matcher) else matcher
    if pattern and not fnmatch.fnmatch(candidate, pattern):
        return format_records([]), IOResult()
    out = p0.original if isinstance(p0, PathSpec) else p0
    return format_records([out]), IOResult()

File: typescript/packages/core/src/core/github/find.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GitHubAccessor } from '../../accessor/github.ts'
import type { FindOptions } from '../../resource/base.ts'
import type { PathSpec } from '../../types.ts'
import { buildTree, keep } from '../../commands/builtin/findEval.ts'
import { stripSlash } from '../../utils/slash.ts'

function strip(path: PathSpec): string {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  return stripSlash(p)
}

export function find(
  accessor: GitHubAccessor,
  path: PathSpec,
  options: FindOptions = {},
): Promise<string[]> {
  const base = strip(path)
  const prefix = base === '' ? '' : `${base}/`
  const baseDepth = base === '' ? 0 : (base.match(/\//g) ?? []).length + 1
  const results: string[] = []
  const tree =
    options.tree ??
    buildTree({
      name: options.name,
      iname: options.iname,
      pathPattern: options.pathPattern,
      type: options.type,
      nameExclude: options.nameExclude,
      orNames: options.orNames,
    })
  const sortedKeys = Object.keys(accessor.tree).sort()
  for (const p of sortedKeys) {
    if (p !== base && !p.startsWith(prefix)) continue
    const entry = accessor.tree[p]
    if (entry === undefined) continue
    const isDir = entry.type === 'tree'
    const fullPath = `/${p}`
    const depth = (p.match(/\//g) ?? []).length + 1 - baseDepth
    if (options.maxDepth !== null && options.maxDepth !== undefined && depth > options.maxDepth) {
      continue
    }
    const entryName = p.split('/').pop() ?? p
    if (
      !keep(
        { key: fullPath, name: entryName, kind: isDir ? 'd' : 'f', depth },
        tree,
        options.minDepth,
      )
    ) {
      continue
    }
    const size = entry.size ?? 0
    if (options.minSize !== null && options.minSize !== undefined && size < options.minSize) {
      continue
    }
    if (options.maxSize !== null && options.maxSize !== undefined && size > options.maxSize) {
      continue
    }
    results.push(fullPath)
  }
  return Promise.resolve(results)
}
